cql3/statements/select_statement.cc

```diff
@@ -992,31 +996,40 @@ indexed_table_select_statement::do_execute(service::storage_proxy& proxy,
             // page size is set to the internal count page size, regardless of the user-provided value
             internal_options.reset(new cql3::query_options(std::move(internal_options), options.get_paging_state(), DEFAULT_COUNT_PAGE_SIZE));
             return repeat([this, &builder, &options, &internal_options, &proxy, &state, now, whole_partitions, partition_slices, restrictions_need_filtering] () {
-                auto consume_results = [this, &builder, &options, &internal_options, restrictions_need_filtering] (foreign_ptr<lw_shared_ptr<query::result>> results, lw_shared_ptr<query::read_command> cmd) {
+                auto consume_results = [this, &builder, &options, &internal_options, &proxy, &state, restrictions_need_filtering] (foreign_ptr<lw_shared_ptr<query::result>> results, lw_shared_ptr<query::read_command> cmd, lw_shared_ptr<const service::pager::paging_state> paging_state) {
+                    if (paging_state) {
+                        paging_state = generate_view_paging_state_from_base_query_results(paging_state, results, proxy, state, options);
```

Comment:
@avelanarius do I understand correctly that this is the heart of this fix (apart from set_ranges issue)? We use paging state computed based on base table paging state instead of index paging state.

What is not 100% clear to me yet is why base table select can return more results than index select?

Reply:
The fix for paging has two parts:
1. Setting `paging_state` on `internal_options` after base query - not before (here we do it in `consume_results`).
2. Computing next `paging_state` based on base table select and index paging state. 

Base select always reads an entire partition, so in case of table:
```
pk | ck1 | ck2
 1 |   1 |   1
 1 |   1 |   2
 1 |   1 |   3
 1 |   1 |   4
...
 1 |   1 |   100000
```
First, the index query (on indexed column `ck1 = 1`), will return a page of results (for example up to `ck2 = 10000`). However, when executing base query we will read an entire partition (`pk = 1`) - more than index. This is why we need to adjust the paging state to avoid reading the rows again (we would read from `ck2 = 10001`).

`generate_view_paging_state_from_base_query_results` was also already used on non-aggregate code path.

I described this paging behaviour in second example in https://github.com/scylladb/scylla/issues/7432.